### ISSUE TITLE
Paper details

### ISSUE THREAD

**Issue by justanhduc** [CompVis/brushstroke-parameterized-style-transfer#1]
Hi authors,
Congratulations on the acceptance! I am trying to reproduce the paper. I have the following questions regarding the details of the paper in order to implement it. 

1. What exactly are the parameters of a stroke? I thought it should be 3 control points (6 params) + RGB (3 params) + width (1 param) but in the paper you mentioned you used 12 params in total. In Figure 12, you said that `d` is used, but isn't it redundant given `p2`?
2. Do you use tanh or sigmoid to control the locations and colors of strokes into `[0,1]`?
3. In Algorithm 1, is the formula of `M_strokes` right? Since the values of the 2-norm are positive, the mask then should be in `[.5,1]` (or `[0, .5]` if the temperature is negative), which is not in the full range `[0,1]` as it should.
4. In Algorithm 1, is `A=softmax(- t.D_strokes, axis=3)`, the negative logit instead of positive as in the paper?
5. What is the learning rate did you use? Did you use any lr scheduler?

THanks in advance!

**Comment by justanhduc** [comment]
Also, when I follow Section C2 in the supplementary to approximate the distance tensor, the stylized images have visible blocking artifacts. Could you please let me know how you dealt with this problem?

**Comment by dimakot55** [comment]
Hey  @justanhduc, thanks for your interest in our work.

1. Two more parameters are brushstroke center coordinates. 3 control points are given relative to the brushstrokes center. The direction (or orientation) vector `d` is used to control the orientation of strokes, see Sec. B and C. This vector is indeed redundant but it's convenient to use it for explanation purposes.
2. At each optimization step we clip the location to be in the range [0; 1]. No need to clip the color values - but it rather depends on particular implementation. We did not see any benefit applying activation functions. 
3. There was a typo in the paper. The correcti expression is  M_{strokes} = max(sigm(t * ( w_{strokes} - D_{strokes} ) , axis=4) .
4. Yes, minus sign is missing there. Actually, you can use not `t` but a different large constant. 
5. We use no learning rate scheduler. Default learning rate for adam is 0.1 but you may have to change it depending on how you prefer your images stylized.

**Comment by dimakot55** [comment]
> Also, when I follow Section C2 in the supplementary to approximate the distance tensor, the stylized images have visible blocking artifacts. Could you please let me know how you dealt with this problem?

Can you provide an image?

**Comment by justanhduc** [comment]
> Hey @justanhduc, thanks for your interest in our work.
> 
>     1. Two more parameters are brushstroke center coordinates. 3 control points are given relative to the brushstrokes center. The direction (or orientation) vector `d` is used to control the orientation of strokes, see Sec. B and C. This vector is indeed redundant but it's convenient to use it for explanation purposes.
> 
>     2. At each optimization step we clip the location to be in the range [0; 1]. No need to clip the color values - but it rather depends on particular implementation. We did not see any benefit applying activation functions.
> 
>     3. There was a typo in the paper. The correcti expression is  M_{strokes} = max(sigm(t * ( w_{strokes} - D_{strokes} ) , axis=4) .
> 
>     4. Yes, minus sign is missing there. Actually, you can use not `t` but a different large constant.
> 
>     5. We use no learning rate scheduler. Default learning rate for adam is 0.1 but you may have to change it depending on how you prefer your images stylized.

Hi @dimakot55! Thanks a lot for your reply. Many things have been cleared up. I still have some questions regarding your answer.

3. In Algorithm 1, D_{strokes} has shape `hxwxn` and w_{strokes} has shape `1x1xn`, then there can't be `max` across axis 4 can it? Or I guess it should be D (shape `hxwxnxs`) rather than D_{strokes} if there is a max across axis 4.
4. What are the temperature values did you use for sigmoid and softmax?

**Comment by justanhduc** [comment]
> > Also, when I follow Section C2 in the supplementary to approximate the distance tensor, the stylized images have visible blocking artifacts. Could you please let me know how you dealt with this problem?
> 
> Can you provide an image?


Please check the image below.

![stroke-stylized_999_0](https://user-images.githubusercontent.com/15206627/114497328-d4708400-9c5c-11eb-9508-32cb9951079a.jpg)

The blocking artifacts come from the nearest neighbor upsampling, as it assigns different sets of strokes to different blocks. I guess there's something fundamentally wrong in my implementation, but I can't find out what it is.
This is a sample of strokes from my code. is there anything wrong with it?
![strokes](https://user-images.githubusercontent.com/15206627/114497213-92dfd900-9c5c-11eb-9511-013727fa5aa2.jpg)

PS: currently I'm running 1000 steps following your paper. The quality can be better if I run more, but the blocking artifacts still exist.

**Comment by matthias-wright** [comment]
Hi @justanhduc, thanks for your interest. Have a look [here](https://github.com/CompVis/brushstroke-parameterized-style-transfer/blob/f35c9a40f8aefe5b63954fbafe5b34810852e8df/ops.py#L57). This is where the rendering happens. Hope that helps!

**Comment by justanhduc** [comment]
Thanks @matthias-wright. I can finally make it right. I made my implementation in Pytorch available [here](https://github.com/CompVis/brushstroke-parameterized-style-transfer/pull/4) just in case you are interested.

**Comment by TZYSJTU** [comment]
> > Hey @justanhduc, thanks for your interest in our work.
> > ```
> > 1. Two more parameters are brushstroke center coordinates. 3 control points are given relative to the brushstrokes center. The direction (or orientation) vector `d` is used to control the orientation of strokes, see Sec. B and C. This vector is indeed redundant but it's convenient to use it for explanation purposes.
> > 
> > 2. At each optimization step we clip the location to be in the range [0; 1]. No need to clip the color values - but it rather depends on particular implementation. We did not see any benefit applying activation functions.
> > 
> > 3. There was a typo in the paper. The correcti expression is  M_{strokes} = max(sigm(t * ( w_{strokes} - D_{strokes} ) , axis=4) .
> > 
> > 4. Yes, minus sign is missing there. Actually, you can use not `t` but a different large constant.
> > 
> > 5. We use no learning rate scheduler. Default learning rate for adam is 0.1 but you may have to change it depending on how you prefer your images stylized.
> > ```
> 
> Hi @dimakot55! Thanks a lot for your reply. Many things have been cleared up. I still have some questions regarding your answer.
> 
> 3. In Algorithm 1, D_{strokes} has shape `hxwxn` and w_{strokes} has shape `1x1xn`, then there can't be `max` across axis 4 can it? Or I guess it should be D (shape `hxwxnxs`) rather than D_{strokes} if there is a max across axis 4.
> 4. What are the temperature values did you use for sigmoid and softmax?
Is Q3 solved?
